js-api-spec/compiler.test.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright 2023 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import {initCompiler, initAsyncCompiler} from 'sass';
+import type {Compiler, AsyncCompiler} from 'sass';
+
+import {sandbox} from './sandbox';
+
+describe('Compiler', () => {
+  let compiler: Compiler;
+
+  beforeEach(() => {
+    compiler = initCompiler();
+  });
+
+  afterEach(() => {
+    compiler.dispose();
+  });
+
+  describe('compileString', () => {
+    it('performs multiple compilations', () => {
+      expect(compiler.compileString('$a: b; c {d: $a}').css).toBe(
+        'c {\n  d: b;\n}'
+      );
+      expect(compiler.compileString('$a: 1; c {d: $a}').css).toBe(
+        'c {\n  d: 1;\n}'
+      );
+    });
+
+    it('throws after being disposed', () => {
+      compiler.dispose();
+      expect(() => compiler.compileString('$a: b; c {d: $a}')).toThrowError();
+    });
+  });
+
+  describe('compile', () => {
+    it('performs multiple compilations', () => {
+      sandbox(dir => {
+        dir.write({'input.scss': '$a: b; c {d: $a}'});
+        dir.write({'input2.scss': '$a: 1; c {d: $a}'});
+        expect(compiler.compile(dir('input.scss')).css).toBe('c {\n  d: b;\n}');
+        expect(compiler.compile(dir('input2.scss')).css).toBe(
+          'c {\n  d: 1;\n}'
+        );
+      });
+    });
+
+    it('throws after being disposed', () => {
+      sandbox(dir => {
+        dir.write({'input.scss': '$a: b; c {d: $a}'});
+        compiler.dispose();
+        expect(() => compiler.compile(dir('input.scss'))).toThrowError();
+      });
+    });
+  });
+});
+
+describe('AsyncCompiler', () => {
+  let compiler: AsyncCompiler;
+
+  beforeEach(async () => {
+    compiler = await initAsyncCompiler();
+  });
+
+  afterEach(async () => {
+    await compiler.dispose();
+  });
+
+  describe('compileStringAsync', () => {
+    it('performs multiple compilations', async () => {
+      expect((await compiler.compileStringAsync('$a: b; c {d: $a}')).css).toBe(
+        'c {\n  d: b;\n}'
+      );
+      expect((await compiler.compileStringAsync('$a: 1; c {d: $a}')).css).toBe(
+        'c {\n  d: 1;\n}'
+      );
+    });
+
+    it('throws after being disposed', async () => {
+      await compiler.dispose();
+      expect(() =>
+        compiler.compileStringAsync('$a: b; c {d: $a}')
+      ).toThrowError();
+    });
+
+    it('waits for compilations to finish before disposing', async () => {
+      const compilation = compiler.compileStringAsync('$a: b; c {d: $a}');
+      await compiler.dispose();
+      await expectAsync(compilation).toBeResolved();
+    });
+  });
+
+  describe('compileAsync', () => {
+    it('performs multiple compilations', async () => {
+      await sandbox(async dir => {
```

Comment:
This call back can be slightly simplified by returning sandbox directly from the arrow function instead of wrapping it in an async function-

```
it('performs multiple compilations', () => 
  sandbox(async dir => {
```

(Side note, if the promise chain isn't quite right with the sandbox, I found it was very easy to have false positives. All of these look good, but I got into the habit of changing an expected output to verify that the test can actually break)

Reply:
Unfortunately, If I don't await for sandbox I get:

```
**************************************************
*                    Failures                    *
**************************************************

1) AsyncCompiler compileAsync performs multiple compilations
  - Unhandled promise rejection: Error: Compiler caused error: Async compiler has already been disposed..

2) Non-spec failure
  - Spec "AsyncCompiler compileAsync waits for compilations to finish before disposing" ran a "toBeResolved" expectation after it finished.
  1. Did you forget to return or await the result of expectAsync?
  2. Was done() invoked before an async operation completed?
  3. Did an expectation follow a call to done()?
```

Reply:
Fixed in 95fd55894208bcd297024c2ba99fafb91cfec8f0